### PR TITLE
[Android] Replace custom StackLock with platform lock

### PR DESCRIPTION
**Pull request by austinh0**
#### Problem
After #9884, Android is using the platform event loop in `GenericPlatformManagerImpl_POSIX`, which uses the platform lock. However, Android is still using its own custom lock, which results in deadlocks.

#### Change overview
* Switch to using platform lock:
  * Remove Android-only `StackLock.h` and custom `JniReferences::GetInstance().GetStackLock()`.
  * Replace `StackLockGuard` with `chip::DeviceLayer::StackLock`, and `StackUnlockGuard` with newly created `chip::DeviceLayer::StackUnlock`

#### Testing
* Commissioned and controlled device using Android CHIPTool.

### CODE REVIEW

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp [hunk 1]
```diff
@@ -56,8 +56,6 @@ System::LayerSocketsLoop & SystemLayerSocketsLoop()
 template <class ImplClass>
 CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_InitChipStack()
 {
-    mChipStackLock = PTHREAD_MUTEX_INITIALIZER;
```
Comment (msandstedt) [inline src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:59]: Why the change to move this out of _InitChipStack?  Now _InitChipStack and _Shutdown are asymmetric and internalization of the stack is impossiblle.

**Review by msandstedt — CHANGES_REQUESTED** [review]
PR breaks POSX::_InitChipStack such that reinitialization of chip stack does not initialize the stack lock.

File: src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp (hunk 1 above)
Comment (austinh0) [reply src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp]: I meant to revert this, thanks for catching.

**Review by msandstedt — APPROVED** [review]

**Comment by austinh0** [comment]
@bzbarsky-apple @chrisdecenzo @Damian-Nordic @jmartinez-silabs @saurabhst?

**Review by jmartinez-silabs — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]
